cloud/amazon/ec2_vpc_dhcp_options.py

```diff
@@ -128,105 +157,204 @@
+            found = True
+            dhcp_option = existing_options
+            module.exit_json(changed=changed, new_options=new_options, dhcp_options_id=existing_options.id)
+
+        # If no vpc_id was given, or the options don't match then look for an existing set using tags
+        found, dhcp_option = match_dhcp_options(connection, params['tags'], new_options)
+
+    # Now let's cover the case where there are existing options that we were told about by id
+    # If a dhcp_options_id was supplied we don't look at options inside, just set tags (if given)
+    else:
+        supplied_options = connection.get_all_dhcp_options(dhcp_options_ids=params['dhcp_options_id'])
+        found = True
+        dhcp_option = supplied_options[0]
+        if params['state'] != 'absent' and options['tags']:
+            ensure_tags(connection, options['tags'], False, module.check_mode)
+
+    # Now we have the dhcp options set, let's do the necessary
+
+    # if we found options we were asked to remove then try to do so
+    if found and params['state'] == 'absent' and not module.check_mode:
+        changed = remove_dhcp_options_by_id(connection, dhcp_option.id)
+        module.exit_json(changed=changed, new_options={})
+
+    # otherwise if we haven't found the required options we have something to do
+    elif not module.check_mode and not found:
+
+        # create some dhcp options if we weren't able to use existing ones
+        if not found:
+            # Convert netbios-node-type and domain-name back to strings
+            if new_options['netbios-node-type']:
+                new_options['netbios-node-type'] = new_options['netbios-node-type'][0]
+            if new_options['domain-name']:
+                new_options['domain-name'] = new_options['domain-name'][0]
+
+            # create the new dhcp options set requested
+            dhcp_option = connection.create_dhcp_options(
+                new_options['domain-name'],
+                new_options['domain-name-servers'],
+                new_options['ntp-servers'],
+                new_options['netbios-name-servers'],
+                new_options['netbios-node-type'])
+            changed = True
+            if params['tags']:
+                ensure_tags(connection, dhcp_option.id, params['tags'], False, module.check_mode)
+
+    # If we were given a vpc_id, then attach the options we now have to that before we finish
+    if params['vpc_id'] and not module.check_mode:
         connection.associate_dhcp_options(dhcp_option.id, params['vpc_id'])
-        if params['delete_old'] and existing_options:
-            other_vpcs = _get_vpcs_by_dhcp_options(existing_options.id, connection)
-            if len(other_vpcs) == 0 or (len(other_vpcs) == 1 and other_vpcs[0].id == params['vpc_id']):
-                connection.delete_dhcp_options(existing_options.id)
+        # and remove old ones if that was requested
+        if params['delete_old'] and params['vpc_id'] and existing_options:
```

Comment:
The check for `params['vpc_id']` is redundant here since it's in the parent if clause.